docs/source/developer_guide/library_design.md

```diff
@@ -0,0 +1,111 @@
+Geoarrow allows a mixture of geometry types to present in the same column by adopting the [Dense Union Array Layout](https://arrow.apache.org/docs/format/Columnar.html#dense-union).
+In cusptial,
+we refer to each level of indirection from highest level to lowest by
+`geometries`, `parts`, `rings` and `coordinates`.
+The first three are integral offset arrays and the last is an floating-point interleaved xy-coordinated array.
+Read [geoarrow format documentation](https://github.com/geopandas/geo-arrow-spec/blob/main/format.md) specification for more detail.
+
+Cuspatial implements partial arrow union array via `GeoColumn` and `GeoMeta`.
+A GeoColumn is a composition of child columns and a `GeoMeta` object.
+The `GeoMeta` owns two arrays that are similar to the types buffer and offsets buffer from dense union.
+
+```{note} Currently,
+`GeoColumn` only implements four concrete array types: `points`, `multipoints`, `multilinestrings` (named as lines) and `multipolygon` (named as polygons).
+Linestrings and multilinestrings are stored uniformly as multilinestrings in the multilinestrings array.
+Polygons and multipolygons are stored uniformly as multipolygons in multipolygon array.
+Points and multipoints are stored separately in different arrays.
+This is because storing points in a multipoints array requires 50% more storage overhead.
+While this may also be true for linestrings and polygons,
+previous use cases of cuspatial often involves more complex linestrings and polygons,
+and the storage overhead is lower compared to points.
+```
+
+`GeoSeries` and `GeoDataFrame` inherits from `cudf.Series` and `cudf.DataFrame` respectively.
+`Series` and `DataFrame` are all generic `Frame` objects which represents a collection of generic columns.
+cuSpatial extends these cudf objects by allowing `GeoColumn` to present in the frame.
+
+`GeoSeries` and `GeoDataFrame` are convertible to and from `geopandas`.
+Interoperability between cuspatial, `geopandas` and other data format is maintained in `cuspatial.io` package.
+
+### UnionArray Compliance
+
+As previously mentioned,
+cuspatial's `GeoColumn` is an specialization of arrow's `UnionArray`.
+A fundamental addition in data types should be implemented in `cudf` and `GeoColumn` should simply inherits its functionality.
+However,
+`UnionArray` stands distinctly from existing data types in libcudf and requires substantial effort to implement.
+In the interim,
+cuspatial developers should build `GeoColumn` complying to `UnionArray` standards.
+Such effort may be upstreamed to `cudf` when development is more underway.
+
+## GIS computation APIs
+
+Besides the data structure,
+cuspatial maintains a set of computation APIs.
+The computation APIs are organized into several modules.
+All spatial computation modules are further grouped into a `spatial` subpackage.
+Developers are encouraged to use specific names for the module of the function added in the PR.
```

Comment:
What do you mean by "Developers are encouraged to use specific names for the module of the
function added in the PR."  What specific names should be used?  Also, it's weird to refer to PRs here in the library design documentation.


```suggestion
## Geospatial computation APIs

In addition to data structures, cuSpatial provides a set of computation APIs.
The computation APIs are organized into several modules. All spatial
computation modules are further grouped into a `spatial` subpackage.
Developers are encouraged to use specific names for the module of the
function added in the PR.
```

Reply:
This is opposed to using names like `generics.py` for module names. We can make it clear in this paragraph:

```suggestion
## Geospatial computation APIs

In addition to data structures, cuSpatial provides a set of computation APIs.
The computation APIs are organized into several modules. All spatial
computation modules are further grouped into a `spatial` subpackage.
Module names should correspond to a specific computation category,
such as `distance` or `join`. Cuspatial avoids using general category names,
such as `generic`.
```